In [18]:
#Extract packages
import pandas as pd
import numpy as np

In [28]:
#Extract dataset
df = pd.read_csv('https://raw.githubusercontent.com/HumayDS/Digital-Data-Analytics-E-15-24/refs/heads/main/clv.csv')

In [29]:
df.head()

,customer_id,first_purchase_date,last_purchase_date,total_orders,total_revenue
0,C001,2023-01-10,2023-03-15,5,250
1,C002,2023-02-05,2023-02-20,2,80
2,C003,2023-01-20,2023-06-10,10,900
3,C004,2023-03-01,2023-03-05,1,30
4,C005,2023-01-15,2023-04-20,6,300


#CLV - Customer lifetime value

##Customer Lifetime Value (CLV) is the total amount of revenue or profit a business expects to earn from a single customer over the entire duration of their relationship.
##CLV answers the question: “How much money will this customer bring to the business over time?”
##CLV is mainly driven by:
##Average Order Value (AOV) – how much a customer spends per purchase
##Purchase Frequency – how often they buy
##Customer Lifespan – how long they stay a customer

#CLV=Average Order Value×Purchase Frequency×Customer Lifespan

Example:

A customer spends $20 per purchase

Buys 2 times per month

Stays for 12 months

CLV = 20 × 2 × 12 = $480

Why CLV matters:

Helps determine how much you can spend on acquiring a customer

Identifies high-value customer segments

Supports retention strategies

Improves profitability analysis

In [30]:
# Recommendable check NA, outlier(describe(), and data types before analysing. )
#check if there is any na
df.isnull().sum()

,0
customer_id,0
first_purchase_date,0
last_purchase_date,0
total_orders,0
total_revenue,0


In [ ]:
#df = df.dropna()

In [32]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
total_orders,60.0,5.016667,3.427168,1.0,2.00,4.5,8.0,12.0
total_revenue,60.0,383.950000,381.303492,20.0,83.75,225.0,645.0,1400.0


In [33]:
# Convert date columns to datetime format
df['first_purchase_date'] = pd.to_datetime(df['first_purchase_date'])
df['last_purchase_date'] = pd.to_datetime(df['last_purchase_date'])

In [35]:
# Calculate customer lifetime in days
# The total number of days the customer has been with us.
df['customer_lifetime'] = (df['last_purchase_date'] - df['first_purchase_date']).dt.days

In [36]:
# Calculate purchase frequency
#The average number of purchases a customer makes per day.
df['purchase_frequency'] = df['total_orders'] / df['customer_lifetime']

df

,customer_id,first_purchase_date,last_purchase_date,total_orders,total_revenue,customer_lifetime,purchase_frequency
0,C001,2023-01-10,2023-03-15,5,250,64,0.078125
1,C002,2023-02-05,2023-02-20,2,80,15,0.133333
2,C003,2023-01-20,2023-06-10,10,900,141,0.070922
3,C004,2023-03-01,2023-03-05,1,30,4,0.250000
4,C005,2023-01-15,2023-04-20,6,300,95,0.063158
5,C006,2023-02-10,2023-02-12,1,20,2,0.500000
6,C007,2023-01-05,2023-07-01,12,1200,177,0.067797
7,C008,2023-03-12,2023-04-01,3,150,20,0.150000
8,C009,2023-02-18,2023-05-30,7,500,101,0.069307
9,C010,2023-01-25,2023-01-30,2,60,5,0.400000


In [37]:
# Calculate average order value
#AOV (Average Order Value) calculates how much a customer spends on average per order.
df['average_order_value'] = df['total_revenue'] / df['total_orders']


In [43]:
segment_counts = df['segment'].value_counts().sort_index()
print(segment_counts)# calculates Customer Lifetime Value (CLV) for each customer.
#This represents the total revenue a business can expect from a customer over their lifetime.
df['CLV'] = (
    df['average_order_value'] *
    df['purchase_frequency'] *
    df['customer_lifetime']
)

df

,customer_id,first_purchase_date,last_purchase_date,total_orders,total_revenue,customer_lifetime,purchase_frequency,average_order_value,CLV,segment
0,C001,2023-01-10,2023-03-15,5,250,64,0.078125,50.000000,250.0,Medium
1,C002,2023-02-05,2023-02-20,2,80,15,0.133333,40.000000,80.0,Low
2,C003,2023-01-20,2023-06-10,10,900,141,0.070922,90.000000,900.0,High
3,C004,2023-03-01,2023-03-05,1,30,4,0.250000,30.000000,30.0,Low
4,C005,2023-01-15,2023-04-20,6,300,95,0.063158,50.000000,300.0,Medium
5,C006,2023-02-10,2023-02-12,1,20,2,0.500000,20.000000,20.0,Low
6,C007,2023-01-05,2023-07-01,12,1200,177,0.067797,100.000000,1200.0,High
7,C008,2023-03-12,2023-04-01,3,150,20,0.150000,50.000000,150.0,Medium
8,C009,2023-02-18,2023-05-30,7,500,101,0.069307,71.428571,500.0,High
9,C010,2023-01-25,2023-01-30,2,60,5,0.400000,30.000000,60.0,Low


In [48]:
segment_counts = df['segment'].value_counts().sort_index()
print(segment_counts)

segment
Low       20
Medium    20
High      20
Name: count, dtype: int64


In [44]:
#This line segments customers into three groups based on their CLV values using quantile-based binning.
df['segment'] = pd.qcut(df['CLV'], q=3, labels=['Low', 'Medium', 'High'])

In [46]:
df

,customer_id,first_purchase_date,last_purchase_date,total_orders,total_revenue,customer_lifetime,purchase_frequency,average_order_value,CLV,segment
0,C001,2023-01-10,2023-03-15,5,250,64,0.078125,50.000000,250.0,Medium
1,C002,2023-02-05,2023-02-20,2,80,15,0.133333,40.000000,80.0,Low
2,C003,2023-01-20,2023-06-10,10,900,141,0.070922,90.000000,900.0,High
3,C004,2023-03-01,2023-03-05,1,30,4,0.250000,30.000000,30.0,Low
4,C005,2023-01-15,2023-04-20,6,300,95,0.063158,50.000000,300.0,Medium
5,C006,2023-02-10,2023-02-12,1,20,2,0.500000,20.000000,20.0,Low
6,C007,2023-01-05,2023-07-01,12,1200,177,0.067797,100.000000,1200.0,High
7,C008,2023-03-12,2023-04-01,3,150,20,0.150000,50.000000,150.0,Medium
8,C009,2023-02-18,2023-05-30,7,500,101,0.069307,71.428571,500.0,High
9,C010,2023-01-25,2023-01-30,2,60,5,0.400000,30.000000,60.0,Low


In [47]:
segment_analysis = df.groupby('segment').agg({
    'CLV': 'mean',
    'total_orders': 'mean',
    'total_revenue': 'mean',
    'customer_lifetime': 'mean',
    'purchase_frequency': 'mean'
}).round(2)

segment_analysis

/tmp/ipykernel_333/930912433.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  segment_analysis = df.groupby('segment').agg({


,CLV,total_orders,total_revenue,customer_lifetime,purchase_frequency
segment,,,,,
Low,58.35,1.60,58.35,11.20,0.17
Medium,242.50,4.25,242.50,65.45,0.08
High,851.00,9.20,851.00,144.50,0.06


#Interesting insight :
##Best customers are not frequent — they are long-term

CLV ↑ increases mainly due to lifetime and the number of orders.

As we move from Low → High:

Orders: 1.6 → 9.2
Lifetime: 11 → 144 days
 This means:
The main driver of CLV growth is not frequency, but longevity + order count.

Interesting and critical finding ❗

Purchase frequency DECREASES:

Low: 0.17
Medium: 0.08
High: 0.06

What does this mean?

High-value customers do not purchase frequently.

However, they stay longer and place more orders over time.
They are loyal, but not impulsive.

### Low CLV

Short-term, few orders

They come quickly and leave quickly
 Retention problem

### Medium CLV

Moderate behavior, stable customers
Upsell / growth potential

###High CLV

Stay for a long time, place many orders

But purchase frequency is low
Loyal customers, but need activation